aptos-move/framework/aptos-token/sources/token.move

```diff
@@ -1245,6 +1281,15 @@ module aptos_token::token {
         }
     }
 
+    public fun get_tokendata_maximum(creator: address, token_data_id: TokenDataId): u64 acquires Collections {
+        assert!(exists<Collections>(creator), error::not_found(ECOLLECTIONS_NOT_PUBLISHED));
+        let all_token_data = &borrow_global<Collections>(creator).token_data;
+        assert!(table::contains(all_token_data, token_data_id), error::not_found(ETOKEN_DATA_NOT_PUBLISHED));
```

Comment:
consolidate these lines into a private function: `assert_token_data_at_address` since you have similar pattern in all the getters.

Reply:
I think the problem is that we'll have to call `let all_token_data = &borrow_global<Collections>(creator).token_data;` twice if we do that? once in the `assert_token_data_at_address` function and later when we're actually mutating the tokendata:
`fun assert_token_data_at_address(creator: &signer, token_data_id: TokenDataId) {
           let creator_addr = token_data_id.creator;
            assert!(signer::address_of(creator) == creator_addr, error::permission_denied(ENO_MUTATE_CAPABILITY));
            assert!(exists<Collections>(creator_addr), error::not_found(ECOLLECTIONS_NOT_PUBLISHED));
            // 1st time getting all token data
            let all_token_data = &mut borrow_global_mut<Collections>(
                creator_addr
            ).token_data;
    
           assert!(table::contains(all_token_data, token_data_id), error::not_found(ETOKEN_DATA_NOT_PUBLISHED));
}`
`public fun mutate_tokendata_royalty(creator: &signer, token_data_id: TokenDataId, royalty: Royalty) acquires Collections {
        assert_token_data_at_address(creator, token_data_id);
        // 2nd time getting all token data
        let all_token_data = &mut borrow_global_mut<Collections>(
                creator_addr
            ).token_data;
        let token_data = table::borrow_mut(all_token_data, token_data_id);
        assert!(token_data.mutability_config.royalty, error::permission_denied(EFIELD_NOT_MUTABLE));
        token_data.royalty = royalty;
    }
`
^ feel like this is an efficient use of gas

I was also thinking about creating a generic function for mutating token data, but that's kind of ugly since we'll have to take in a lot of optional arguments... for example: 
`fun mutate_tokendata_generic(
    creator: &signer,
    token_data_id: TokenDataId,
    maximum: Option<u64>,
    uri: Option<String>,
    royalty: Option<Royalty>,
    description: Option<String>,
    keys: Option<vector<String>>,
    values: Option<vector<vector<u8>>>,
    types: Option<vector<String>>,
) acquires Collections {
    let creator_addr = token_data_id.creator;
    assert!(signer::address_of(creator) == creator_addr, ENO_MUTATE_CAPABILITY);
    assert!(exists<Collections>(creator_addr), ECOLLECTIONS_NOT_PUBLISHED);
    let all_token_data = &mut borrow_global_mut<Collections>(
        creator_addr
    ).token_data;

    assert!(table::contains(all_token_data, token_data_id), error::not_found(ETOKEN_DATA_NOT_PUBLISHED));
    let token_data = table::borrow_mut(all_token_data, token_data_id);

    if (option::is_some(&uri)) {
        let uri_value = option::extract(&mut uri);
        assert!(string::length(&uri_value) <= MAX_URI_LENGTH, error::invalid_argument(EURI_TOO_LONG));
        assert!(token_data.mutability_config.uri, error::permission_denied(EFIELD_NOT_MUTABLE));
        token_data.uri = uri_value;
    };
    if (option::is_some(&maximum)) {
        assert!(token_data.mutability_config.maximum, error::permission_denied(EFIELD_NOT_MUTABLE));
        token_data.maximum = option::extract(&mut maximum);
    };
    if (option::is_some(&royalty)) {
        assert!(token_data.mutability_config.royalty, error::permission_denied(EFIELD_NOT_MUTABLE));
        token_data.royalty = option::extract(&mut royalty);
    };
    if (option::is_some(&description)) {
        assert!(token_data.mutability_config.description, error::permission_denied(EFIELD_NOT_MUTABLE));
        token_data.description = option::extract(&mut description);
    };
    if (option::is_some(&keys) && option::is_some(&values) && option::is_some(&types)) {
        assert!(token_data.mutability_config.properties, error::permission_denied(EFIELD_NOT_MUTABLE));
        property_map::update_property_map(&mut token_data.default_properties, option::extract(&mut keys), option::extract(&mut values), option::extract(&mut types));
    };
}
`
so yeah I think even though the current approach is a bit repetitive, but it's straightforward and cost-efficient.

Reply:
what I mean is put these lines into an `assert_xxx` function since you are calling it in multiple functions. This does not include the `global_borrow` call.
```
        let creator_addr = token_data_id.creator;
        assert!(signer::address_of(creator) == creator_addr, error::permission_denied(ENO_MUTATE_CAPABILITY));
        assert!(exists<Collections>(creator_addr), error::not_found(ECOLLECTIONS_NOT_PUBLISHED));
```
IMO, `generic function` is not a good practice in most cases so I never mean that to simplify our code.

Reply:
oh i was j thinking that going from 3 lines to 1 line is not that different...